lib/fizzy/parser.cpp

```diff
@@ -50,28 +50,34 @@ inline parser_result<uint8_t> parse(const uint8_t* pos, const uint8_t* end)
 }
 */
 
-template <>
-parser_result<ValType> parse(const uint8_t* pos, const uint8_t* end)
+ValType validate_valtype(uint8_t byte)
 {
-    if (pos == end)
-        throw parser_error{"Unexpected EOF"};
-
-    const auto b = *pos++;
-    switch (b)
+    switch (byte)
     {
     case 0x7F:
-        return {ValType::i32, pos};
+        return ValType::i32;
     case 0x7E:
-        return {ValType::i64, pos};
+        return ValType::i64;
     case 0x7D:  // f32
-        return {ValType::f32, pos};
+        return ValType::f32;
     case 0x7C:  // f64
-        return {ValType::f64, pos};
+        return ValType::f64;
     default:
-        throw parser_error{"invalid valtype " + std::to_string(b)};
+        throw parser_error{"invalid valtype " + std::to_string(byte)};
     }
 }
 
+template <>
+parser_result<ValType> parse(const uint8_t* pos, const uint8_t* end)
+{
+    if (pos == end)
+        throw parser_error{"Unexpected EOF"};
+
+    const auto byte = *pos++;
+    const auto valtype = validate_valtype(byte);
+    return {valtype, pos};
+}
+
 const uint8_t* validate_valtype(const uint8_t* pos, const uint8_t* end)
```

Comment:
Any point keeping both this and the template?

Reply:
The template is not available in `parser_expr.cpp`.

Reply:
Yes, but do we need the template?

Reply:
Template is required for `vec<ValType>`.